File: typescript/core/src/heartbeat-connection.ts
import { connectionPoolOf, type ConnectionPool } from "./connection-pool.js";
import type { Queryable } from "./types.js";

/**
 * The smallest pool a worker reserves a heartbeat connection from. The pool must also hold the
 * notification listener and at least one claim, or the reservation would starve dispatch.
 */
const MINIMUM_RESERVING_CAPACITY = 3;

interface ReservedClient extends Queryable {
  release(error?: Error | boolean): void;
  on?(event: "error", listener: (error: Error) => void): unknown;
  removeListener?(event: "error", listener: (error: Error) => void): unknown;
}

/**
 * Why a worker cannot hold one pooled connection for its heartbeats, or undefined when it can.
 *
 * The capacity must be known: a single node-postgres `Client` also has `connect()`, but it cannot
 * lend out a second session.
 */
export function heartbeatReservationProblem(database: Queryable): string | undefined {
  const pool = connectionPoolOf(database);
  if (pool === undefined) return "the queue's database has no connect() and no attached pool";
  const capacity = pool.options?.max;
  if (capacity === undefined) return "the pool's size is unknown";
  if (capacity < MINIMUM_RESERVING_CAPACITY) {
    return `the pool allows ${capacity} connection${capacity === 1 ? "" : "s"} and needs at least ${MINIMUM_RESERVING_CAPACITY}`;
  }
  return undefined;
}

/** Whether a worker can hold one pooled connection for its heartbeats. */
export function canReserveHeartbeatConnection(database: Queryable): boolean {
  return heartbeatReservationProblem(database) === undefined;
}

async function connectClient(pool: ConnectionPool): Promise<ReservedClient> {
  const client = await pool.connect();
  if (
    typeof client !== "object" ||
    client === null ||
    !("query" in client) ||
    !("release" in client)
  ) {
    throw new TypeError("Database connect() did not return a releasable client");
  }
  return client as ReservedClient;
}

/**
 * One pooled connection kept for heartbeats, with every round bounded in time.
 *
 * Handlers that hold every other pooled connection cannot delay a heartbeat queued behind them,
 * because the heartbeat never waits for the shared pool. A round that outlives its bound destroys
 * the connection, which is the client-side cancel: no session setting is involved, so the
 * connection stays safe behind a transaction-mode pooler. The next round reconnects.
 *
 * A connection that fails between rounds, as when PostgreSQL terminates its backend, is discarded
 * the same way. The failure changes no lease: each attempt's watchdog alone decides when one lapses.
 */
export class ReservedConnection {
  private client: Promise<ReservedClient> | undefined;
  private readonly rounds = new Set<Promise<unknown>>();
  private closed = false;

  constructor(private readonly pool: ConnectionPool) {}

  /** Take the connection now, ahead of any handler that could exhaust the pool. */
  reserve(): void {
    if (this.closed) return;
    void this.acquire().catch(() => undefined);
  }

  run<T>(operation: (client: Queryable) => Promise<T>, timeoutMs: number): Promise<T> {
    const round = this.runBounded(operation, timeoutMs);
    this.rounds.add(round);
    void round.then(
      () => this.rounds.delete(round),
      () => this.rounds.delete(round),
    );
    return round;
  }

  /** Stop reserving and return the client to its pool once every round in flight settles. */
  async close(): Promise<void> {
    this.closed = true;
    await Promise.allSettled(this.rounds);
    const pending = this.client;
    this.client = undefined;
    const client = await pending?.catch(() => undefined);
    client?.release();
  }

  private acquire(): Promise<ReservedClient> {
    if (this.closed) {
      return Promise.reject(new Error("The reserved heartbeat connection is closed"));
    }
    if (this.client === undefined) {
      const pending: Promise<ReservedClient> = connectClient(this.pool).then((client) =>
        this.watch(client, pending),
      );
      this.client = pending;
      pending.catch(() => {
        if (this.client === pending) this.client = undefined;
      });
    }
    return this.client;
  }

  /**
   * Discard the client when it reports an error outside a round. A checked-out node-postgres
   * client has no error listener of its own, and an unobserved error event ends the process.
   */
  private watch(client: ReservedClient, pending: Promise<ReservedClient>): ReservedClient {
    if (client.on === undefined) return client;
    const onError = (error: Error): void => this.discard(pending, error);
    client.on("error", onError);
    const release = client.release.bind(client);
    // The listener stays until release, because the connection can still fail before then.
    client.release = (...error) => {
      release(...error);
      client.removeListener?.("error", onError);
    };
    return client;
  }

  private async runBounded<T>(
    operation: (client: Queryable) => Promise<T>,
    timeoutMs: number,
  ): Promise<T> {
    const pending = this.acquire();
    let expired = false;
    let timer: NodeJS.Timeout | undefined;
    const timedOut = new Promise<never>((_resolve, reject) => {
      timer = setTimeout(() => {
        expired = true;
        reject(new Error(`Heartbeat did not finish within ${timeoutMs} ms`));
      }, timeoutMs);
      timer.unref();
    });
    // A client lent after the bound passed is only destroyed; its renewal would arrive too late.
    const renewal = pending.then((client) => (expired ? timedOut : operation(client)));
    try {
      return await Promise.race([renewal, timedOut]);
    } catch (error) {
      this.discard(pending, error);
      throw error;
    } finally {
      clearTimeout(timer);
    }
  }

  private discard(pending: Promise<ReservedClient>, error: unknown): void {
    if (this.client !== pending) return;
    this.client = undefined;
    // Releasing with an error destroys the connection instead of returning it to the pool, which
    // also ends a statement still running on it. A connect that lands later is destroyed as well.
    void pending.then(
      (client) => client.release(error instanceof Error ? error : true),
      () => undefined,
    );
  }
}

/** One worker's hold on the heartbeat connection its pool shares. */
export interface HeartbeatConnectionLease {
  reserve(): void;
  run<T>(operation: (client: Queryable) => Promise<T>, timeoutMs: number): Promise<T>;
  close(): Promise<void>;
}

const sharedConnections = new WeakMap<
  object,
  { connection: ReservedConnection; holders: number }
>();

/**
 * Hold the heartbeat connection shared by every worker on one pool, or undefined when the pool
 * cannot lend one.
 *
 * Workers share it the way they share the notification listener, so a pool gives up one
 * connection for heartbeats however many workers run on it. The last holder to close returns it.
 */
export function holdHeartbeatConnection(database: Queryable): HeartbeatConnectionLease | undefined {
  const pool = connectionPoolOf(database);
  if (pool === undefined || heartbeatReservationProblem(database) !== undefined) return undefined;
  const identity = pool;
  let shared = sharedConnections.get(identity);
  if (shared === undefined) {
    shared = { connection: new ReservedConnection(pool), holders: 0 };
    sharedConnections.set(identity, shared);
  }
  const held = shared;
  held.holders += 1;
  let closing: Promise<void> | undefined;
  return {
    reserve: () => held.connection.reserve(),
    run: (operation, timeoutMs) => held.connection.run(operation, timeoutMs),
    close() {
      closing ??= (async () => {
        held.holders -= 1;
        if (held.holders > 0) return;
        if (sharedConnections.get(identity) === held) sharedConnections.delete(identity);
        await held.connection.close();
      })();
      return closing;
    },
  };
}

File: go/heartbeat_connection.go
package workhorse

import (
	"context"
	"errors"
	"sync"
	"time"

	"github.com/jackc/pgx/v5/pgxpool"
)

// heartbeatConnection keeps one pooled connection for heartbeats, with every round bounded in time.
//
// Handlers that hold every other pooled connection cannot delay a heartbeat queued behind them,
// because the heartbeat never waits for the shared pool. A round that outlives its bound closes the
// connection, so the pool discards it instead of lending it again, and closing also ends the
// statement still running on it. The next round takes a fresh connection. The connection holds no
// session state, so it stays safe behind a transaction-mode pooler.
type heartbeatConnection struct {
	pool *pgxpool.Pool
	// slot serializes rounds, because one PostgreSQL session runs one statement at a time and
	// every worker on this pool sends its rounds here. A round waits for it only until its bound.
	slot   chan struct{}
	mu     sync.Mutex
	conn   *pgxpool.Conn
	rounds int
	closed bool
}

func newHeartbeatConnection(pool *pgxpool.Pool) *heartbeatConnection {
	return &heartbeatConnection{pool: pool, slot: make(chan struct{}, 1)}
}

// reserve takes the connection before any handler can exhaust the pool. A pool that cannot lend one
// now is asked again by the first round.
func (connection *heartbeatConnection) reserve(ctx context.Context, bound time.Duration) {
	reserveContext, cancel := context.WithTimeout(ctx, bound)
	defer cancel()
	_, _ = connection.acquire(reserveContext)
}

// run executes one bounded round on the reserved connection. A round that fails discards the
// connection, so the next round starts from a healthy one.
func (connection *heartbeatConnection) run(
	ctx context.Context,
	bound time.Duration,
	round func(context.Context, Executor) error,
) error {
	roundContext, cancel := context.WithTimeout(ctx, bound)
	defer cancel()
	select {
	case connection.slot <- struct{}{}:
	case <-roundContext.Done():
		return roundContext.Err()
	}
	defer func() { <-connection.slot }()
	if err := connection.startRound(); err != nil {
		return err
	}
	defer connection.finishRound()
	held, err := connection.acquire(roundContext)
	if err != nil {
		return err
	}
	if err := round(roundContext, NewPGXExecutor(held.Conn())); err != nil {
		connection.discard(held)
		return err
	}
	return nil
}

// release stops reserving. A round still in flight owns the connection until it returns, so
// releasing never takes a session out from under a statement.
func (connection *heartbeatConnection) release() {
	connection.mu.Lock()
	connection.closed = true
	if connection.rounds > 0 {
		connection.mu.Unlock()
		return
	}
	held := connection.conn
	connection.conn = nil
	connection.mu.Unlock()
	if held != nil {
		held.Release()
	}
}

func (connection *heartbeatConnection) startRound() error {
	connection.mu.Lock()
	defer connection.mu.Unlock()
	if connection.closed {
		return errors.New(heartbeatConnectionClosedMessage)
	}
	connection.rounds++
	return nil
}

func (connection *heartbeatConnection) finishRound() {
	connection.mu.Lock()
	connection.rounds--
	if connection.rounds > 0 || !connection.closed {
		connection.mu.Unlock()
		return
	}
	held := connection.conn
	connection.conn = nil
	connection.mu.Unlock()
	if held != nil {
		held.Release()
	}
}

func (connection *heartbeatConnection) acquire(ctx context.Context) (*pgxpool.Conn, error) {
	connection.mu.Lock()
	held := connection.conn
	closed := connection.closed
	connection.mu.Unlock()
	if held != nil {
		return held, nil
	}
	if closed {
		return nil, errors.New(heartbeatConnectionClosedMessage)
	}
	acquired, err := connection.pool.Acquire(ctx)
	if err != nil {
		return nil, err
	}
	connection.mu.Lock()
	if connection.conn == nil && !connection.closed {
		connection.conn = acquired
		connection.mu.Unlock()
		return acquired, nil
	}
	held = connection.conn
	connection.mu.Unlock()
	acquired.Release()
	if held == nil {
		return nil, errors.New(heartbeatConnectionClosedMessage)
	}
	return held, nil
}

// discard closes the session so the pool cannot lend it again, which also ends a statement still
// running on it.
func (connection *heartbeatConnection) discard(held *pgxpool.Conn) {
	connection.mu.Lock()
	if connection.conn != held {
		connection.mu.Unlock()
		return
	}
	connection.conn = nil
	connection.mu.Unlock()
	_ = held.Conn().Close(context.Background())
	held.Release()
}

// heartbeatConnectionLease is one worker's hold on the heartbeat connection its pool shares.
type heartbeatConnectionLease struct {
	pool       *pgxpool.Pool
	connection *heartbeatConnection
	released   bool
}

func (lease *heartbeatConnectionLease) reserve(ctx context.Context, bound time.Duration) {
	lease.connection.reserve(ctx, bound)
}

func (lease *heartbeatConnectionLease) run(
	ctx context.Context,
	bound time.Duration,
	round func(context.Context, Executor) error,
) error {
	return lease.connection.run(ctx, bound, round)
}

var heldHeartbeatConnections = struct {
	mu   sync.Mutex
	held map[*pgxpool.Pool]*sharedHeartbeatConnection
}{held: make(map[*pgxpool.Pool]*sharedHeartbeatConnection)}

type sharedHeartbeatConnection struct {
	connection *heartbeatConnection
	holders    int
}

// holdHeartbeatConnection hands the caller the heartbeat connection shared by every worker on one
// pool. Workers share it the way they share the notification listener, so a pool gives up one
// connection for heartbeats however many workers run on it. The last holder to release returns it.
func holdHeartbeatConnection(pool *pgxpool.Pool) *heartbeatConnectionLease {
	heldHeartbeatConnections.mu.Lock()
	defer heldHeartbeatConnections.mu.Unlock()
	shared, ok := heldHeartbeatConnections.held[pool]
	if !ok {
		shared = &sharedHeartbeatConnection{connection: newHeartbeatConnection(pool)}
		heldHeartbeatConnections.held[pool] = shared
	}
	shared.holders++
	return &heartbeatConnectionLease{pool: pool, connection: shared.connection}
}

// release drops this worker's hold. Releasing twice is a no-op, so a worker releases at the end of
// every run and takes a new hold when it runs again.
func (lease *heartbeatConnectionLease) release() {
	heldHeartbeatConnections.mu.Lock()
	if lease.released {
		heldHeartbeatConnections.mu.Unlock()
		return
	}
	lease.released = true
	shared, ok := heldHeartbeatConnections.held[lease.pool]
	if !ok || shared.connection != lease.connection {
		heldHeartbeatConnections.mu.Unlock()
		return
	}
	shared.holders--
	if shared.holders > 0 {
		heldHeartbeatConnections.mu.Unlock()
		return
	}
	delete(heldHeartbeatConnections.held, lease.pool)
	heldHeartbeatConnections.mu.Unlock()
	shared.connection.release()
}
